Make debian package setup example more self explained

6543 opened the pull request:
As most users just copy the example code and the wonder why it did not work ... befor even thinking the shell vars in the example need to be filled with the info below ... just show them. and if there is no selection just return the only options directly

## multible
<img width="1297" height="1020" alt="image" src="https://github.com/user-attachments/assets/7ccd527e-0201-433d-b6ca-178ca25c8faa" />


## only one option available
<img width="1330" height="861" alt="image" src="https://github.com/user-attachments/assets/7d9a2361-42ef-4b8c-a637-cf0eb7e2cbb0" />

## mixed
<img width="1322" height="956" alt="image" src="https://github.com/user-attachments/assets/2bc87bd1-75b9-4e9c-8247-415fb54dd827" />


close https://github.com/go-gitea/gitea/issues/36992

6543 committed:
Make debian package setup example more self explained

silverwind commented:
I'd make the variables uppercase, this is a common practice in shell scripts. For the variables I would do:

```
DISTRIBUTION="alpine" # or "other", "another"
COMPONENT="edge" # or "other", "another"
```

The benefit is the script will work when copied, and it should be obvious what the options are.

6543 commented on templates/package/content/debian.tmpl:
```diff
@@ -4,8 +4,14 @@
+component="" # {{StringUtils.Join .Components ", "}}
+
+{{end}}sudo curl <origin-url data-url="{{AppSubUrl}}/api/packages/{{$.PackageDescriptor.Owner.Name}}/debian/repository.key"></origin-url> -o /etc/apt/keyrings/gitea-{{$.PackageDescriptor.Owner.Name}}.asc
+echo "deb [signed-by=/etc/apt/keyrings/gitea-{{$.PackageDescriptor.Owner.Name}}.asc] <origin-url data-url="{{AppSubUrl}}/api/packages/{{$.PackageDescriptor.Owner.Name}}/debian"></origin-url> {{ if eq (len .Distributions) 1 }}{{ index .Distributions 0 }}{{else}}$distribution{{end}} {{ if eq (len .Components) 1 }}{{ index .Components 0 }}{{else}}$component{{end}}" | sudo tee -a /etc/apt/sources.lis ...
```
```suggestion
echo "deb [signed-by=/etc/apt/keyrings/gitea-{{$.PackageDescriptor.Owner.Name}}.asc] <origin-url data-url="{{AppSubUrl}}/api/packages/{{$.PackageDescriptor.Owner.Name}}/debian"></origin-url> {{ if eq (len .Distributions) 1 }}{{ index .Distributions 0 }}{{else}}$DISTRIBUTION{{end}} {{ if eq (len .Components) 1 }}{{ index .Components 0 }}{{else}}$component{{end}}" | sudo tee -a /etc/apt/sources.list.d/gitea.list
```

6543 committed:
Apply suggestions from code review

Signed-off-by: 6543 <6543@obermui.de>

6543 commented on templates/package/content/debian.tmpl:
```diff
@@ -4,8 +4,14 @@
+component="" # {{StringUtils.Join .Components ", "}}
+
+{{end}}sudo curl <origin-url data-url="{{AppSubUrl}}/api/packages/{{$.PackageDescriptor.Owner.Name}}/debian/repository.key"></origin-url> -o /etc/apt/keyrings/gitea-{{$.PackageDescriptor.Owner.Name}}.asc
+echo "deb [signed-by=/etc/apt/keyrings/gitea-{{$.PackageDescriptor.Owner.Name}}.asc] <origin-url data-url="{{AppSubUrl}}/api/packages/{{$.PackageDescriptor.Owner.Name}}/debian"></origin-url> {{ if eq (len .Distributions) 1 }}{{ index .Distributions 0 }}{{else}}$DISTRIBUTION{{end}} {{ if eq (len .Components) 1 }}{{ index .Components 0 }}{{else}}$component{{end}}" | sudo tee -a /etc/apt/sources.lis ...
```
```suggestion
echo "deb [signed-by=/etc/apt/keyrings/gitea-{{$.PackageDescriptor.Owner.Name}}.asc] <origin-url data-url="{{AppSubUrl}}/api/packages/{{$.PackageDescriptor.Owner.Name}}/debian"></origin-url> {{ if eq (len .Distributions) 1 }}{{ index .Distributions 0 }}{{else}}$DISTRIBUTION{{end}} {{ if eq (len .Components) 1 }}{{ index .Components 0 }}{{else}}$COMPONENT{{end}}" | sudo tee -a /etc/apt/sources.list.d/gitea.list
```

6543 committed:
Apply suggestions from code review

Signed-off-by: 6543 <6543@obermui.de>

6543 committed:
Merge branch 'main' into ux-package-deb-explain-better

silverwind commented on templates/package/content/debian.tmpl:
```diff
@@ -4,8 +4,14 @@
 				<label>{{svg "octicon-terminal"}} {{ctx.Locale.Tr "packages.debian.registry"}}</label>
-				<div class="markup"><pre class="code-block"><code>sudo curl <origin-url data-url="{{AppSubUrl}}/api/packages/{{$.PackageDescriptor.Owner.Name}}/debian/repository.key"></origin-url> -o /etc/apt/keyrings/gitea-{{$.PackageDescriptor.Owner.Name}}.asc
-echo "deb [signed-by=/etc/apt/keyrings/gitea-{{$.PackageDescriptor.Owner.Name}}.asc] <origin-url data-url="{{AppSubUrl}}/api/packages/{{$.PackageDescriptor.Owner.Name}}/debian"></origin-url> $distribution $component" | sudo tee -a /etc/apt/sources.list.d/gitea.list
+				<div class="markup"><pre class="code-block"><code>{{ if gt (len .Distributions) 1 }}# please select one:
+DISTRIBUTION="" # {{StringUtils.Join .Distributions ", "}}
+
+{{end}}{{ if gt (len .Components) 1 }}# please select one:
+COMPONENT="" # {{StringUtils.Join .Components ", "}}
```
```suggestion
        <div class="markup"><pre class="code-block"><code>{{ if gt (len .Distributions) 1 }}DISTRIBUTION="" # one of {{StringUtils.Join .Distributions ", "}}{{end}}
          {{ if gt (len .Components) 1 }}COMPONENT="" # one of {{StringUtils.Join .Components ", "}}
```

6543 committed:
Update templates/package/content/debian.tmpl

Co-authored-by: silverwind <me@silverwind.io>
Signed-off-by: 6543 <6543@obermui.de>

6543 committed:
Merge branch 'main' into ux-package-deb-explain-better

silverwind committed:
make fmt

6543 committed:
Merge branch 'main' into ux-package-deb-explain-better

6543 committed:
With translation

6543 committed:
did not syntaxhiglight so rm it again

6543 commented:
@silverwind you might wana re-review as some bits changed:
- tryed syntax highliting - did not work (reverted)
- fixed intendation
- removed and added translation

6543 commented:
-> screenshots updated to current pull state

silverwind commented:
Looks ok, I would add a empty line after the `COMPONENT` line, just to visually distinguish variables from commands.

wxiaoguang commented:
> Looks ok, I would add a empty line after the `COMPONENT` line, just to visually distinguish variables from commands.

Maybe it's even better to generate the texts from backend:

* It's easier to maintain than template (the template syntax is hard to read the the logic is complex or we need some special formats)
* It can add some tests to see what we want for various cases.

6543 committed:
Merge branch 'main' into ux-package-deb-explain-better

6543 committed:
fmt

6543 commented:
as per if we do it in golang - can we just do it as is ...

6543 commented:
-> screenshots updated

6543 committed:
Merge branch 'main' into ux-package-deb-explain-better

6543 committed:
Merge branch 'main' into ux-package-deb-explain-better

silverwind commented:
> It would be better to generate this information using a backend Go function, as the logic is a bit complex. This approach makes the code easier to maintain and test.

Maybe for later, this is still good to merge as a standalone improvement but I'll check the generation option with Claude.

silverwind committed:
Merge remote-tracking branch 'origin/main' into ux-package-deb-explain-better

* origin/main: (645 commits)
  When the requested arch rpm is missing fall back to noarch (#37236)
  Fix `relative-time` error and improve global error handler (#37241)
  Enhance styling in actions page (#37323)
  fix(oauth): Error on auth sources with spaces (#37327)
  Fix actions concurrency groups cross-branch leak (#37311)
  Fix bug when accessing user badges (#37321)
  Fix AppFullLink (#37325)
  Update go js dependencies (#37312)
  Update GitHub Actions to latest major versions (#37313)
  Revert "Add WebKit to e2e test matrix (#37298)" (#37315)
  Add `form-fetch-action` to some forms, fix "fetch action" resp bug (#37305)
  Move heatmap to first-party code (#37262)
  Use updated yaml fields for snapcraft (#37318)
  Remove dead code identified by `deadcode` tool (#37271)
  Enable strict TypeScript, add `errorMessage` helper (#37292)
  Fix vite manifest update masking build errors (#37279)
  bump snapcraft base (#37301)
  Add WebKit to e2e test matrix (#37298)
  Don't add useless labels which will bother changelog generation (#37267)
  Fix Repository transferring page (#37277)
  ...

# Conflicts:
#	options/locale/locale_en-US.ini
#	templates/package/content/debian.tmpl

6543 commented:
I still think it is best to be maintained in template...

If it's about tests via go ... well in this case we should drop templates all together ... witch makes it harder for user to customize and we just write our own template engine in the end

wxiaoguang commented:
> > It would be better to generate this information using a backend Go function, as the logic is a bit complex. This approach makes the code easier to maintain and test.
> 
> Maybe for later, this is still good to merge as a standalone improvement but I'll check the generation option with Claude.

-> Better Installation Instructions For Debian Packages #36992

> My suggestion is to add drop-down menus above the script for the distribution and component. The values of these menus would then be used to fill in the generated script.

silverwind commented:
Yes, the main problem with having this generated entirely in go is users can not override with custom template. Could be solved via template helper, though.

TheFox0x7 commented:
If you ask me I think it would be a feature :D
Currently if template changes user needs to look and resolve issue - and at times end up reporting it to us, forgetting about the override.

Just the migration would be a large task which we already talked about in templ issue. Reviewing it would be a huge issue

silverwind commented:
Template customization is generally a very bad interface, it requires constant maintenance unless scoped specifically to the block. Much better would be configuration in app.ini.

wxiaoguang commented:
TBH, I think there is a "feature requirement" for improving all the packages related "information" (commands, configurations, etc).

But I don't see it is related to user-side customization, or template rendering engine like templ, or configuration options.


We just need a well-designed "packages information" system to render all useful information in a testable and controllable way, instead of keeping patching the templates.

lunny commented:
> I still think it is best to be maintained in template...
> 
> If it's about tests via go ... well in this case we should drop templates all together ... witch makes it harder for user to customize and we just write our own template engine in the end

Would anyone actually want to customize the setup example? I can't really imagine why this would be needed?

silverwind commented:
Rather unlikely, on second thought. If customization is really needed, it should be provided via config, not crude template replacement.

wxiaoguang commented:
> Rather unlikely, on second thought. If customization is really needed, it should be provided via config, not crude template replacement.

https://github.com/go-gitea/gitea/pull/35564#issuecomment-4288825943

**But I don't see it is related to user-side customization, or template rendering engine like templ, or configuration options.**

6543 commented:
well in this case i must agree that it is unlikely that an user would change that template

silverwind commented:
I'd say merge this now as a strict improvement, then later think about how to improve further.

silverwind committed:
Merge branch 'main' into ux-package-deb-explain-better

lunny commented:
It at least needs a better format.

6543 commented:
> ... better format.

please suggest one, it is to vague for me :upside_down_face:

6543 committed:
Merge branch 'main' into ux-package-deb-explain-better

lunny commented:
I mean like this, and it seems the line `{{if or (gt (len .Distributions) 1) (gt (len .Components) 1)}}{{end -}}` is wrong?
```
<div class="markup">
					<pre class="code-block">
						<code>
						{{- if gt (len .Distributions) 1}}DISTRIBUTION="" # {{ctx.Locale.Tr "packages.debian.registry.one_of_vars"}} {{StringUtils.Join .Distributions ", "}}{{end -}}
						{{- if gt (len .Components) 1}}COMPONENT="" # {{ctx.Locale.Tr "packages.debian.registry.one_of_vars"}} {{StringUtils.Join .Components ", "}}{{end -}}
						{{if or (gt (len .Distributions) 1) (gt (len .Components) 1)}}{{end -}}
						sudo curl {{ctx.AppFullLink}}/api/packages/{{$.PackageDescriptor.Owner.Name}}/debian/repository.key -o /etc/apt/keyrings/gitea-{{$.PackageDescriptor.Owner.Name}}.asc
echo "deb [signed-by=/etc/apt/keyrings/gitea-{{$.PackageDescriptor.Owner.Name}}.asc] {{ctx.AppFullLink}}/api/packages/{{$.PackageDescriptor.Owner.Name}}/debian {{if eq (len .Distributions) 1}}{{index .Distributions 0}}{{else}}$DISTRIBUTION{{end}} {{if eq (len .Components) 1}}{{index .Components 0}}{{else}}$COMPONENT{{end}}" | sudo tee -a /etc/apt/sources.list.d/gitea.list
sudo apt update
						</code>
					</pre>
				</div>
```

6543 commented:
the main probleme here is, tjat we are inside a codeblock and newlines would alter the text :/

wxiaoguang commented:
> the main probleme here is, tjat we are inside a codeblock and newlines would alter the text :/

Yes, so isn't it just the problem that we are discussing? https://github.com/go-gitea/gitea/pull/35564#issuecomment-3375900125

6543 committed:
Merge branch 'main' into ux-package-deb-explain-better

wxiaoguang commented:
Well, just stale again.

Since I proposed a solution and I don't want to see unmaintainable code, so: enhance: user-friendly packages setup manual - #38946
